aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -0,0 +1,160 @@
+#include "ATen/NativeFunctions.h"
+#include "ATen/ATen.h"
+#include <algorithm>
+#include <sstream>
+
+#include "ATen/cuda/AccumulateType.cuh"
+#include "ATen/cuda/CUDATensorMethods.cuh"
+#include "ATen/cuda/CUDATypeConversion.cuh"
+
+
+namespace at {
+namespace native {
+
+// Map the index of an element in tensor from 1D to nD
+__device__ __forceinline__
+void oneD_to_nD(int64_t oneD_index, int64_t* shape_size, int64_t shape_len, int64_t* nD_index) {
+  int64_t res = oneD_index;
+  for (int i = 0; i < shape_len; i++) {
+    int64_t nD_i = oneD_index * shape_len + i;
+    nD_index[nD_i] = res / shape_size[i];
+    res = res % shape_size[i];
+  }
+}
+
+/*
+Map the index of an element in tensor from nD to 1D. A tensor is originally in nD shape,
+and 1D is the unfolded version of it (a vector).
+
+Example: given a 3D tensor
+[
+  [ [1,2], [3,4] ],
+  [ [5,6], [7,8] ]
+]
+
+Here element 3 has nD index = (0,1,0), and this corresponds to oneD index = 2
+*/
+__device__ __forceinline__
+int64_t nD_to_oneD(int64_t* nD_index, int64_t shape_len, int64_t* shape_size, int64_t src_oneD_index) {
```

Comment:
nit: the function naming and variable names are a little strange to me because of the combination of lowercase and upper case characters. It might be easier to refer to "nD_index" as just "indices" and the "oneD_index" as a "linear_index".

How about the following renames:
```
nD_index -> indices
src_oneD_Index -> linear_index
nD_to_oneD -> indices_to_linear_index
shape_len -> dims
```